File: src/Utils/Validator.hs
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Redundant bracket" #-}
{-# HLINT ignore "Use null" #-}
module Utils.Validator where

import Controllers.MatchesController
import Controllers.BoardController
import Controllers.PlayerController
import Controllers.AccountsController
import Data.Char
import Controllers.LettersController
import Utils.Utils as UT

-- Encapsula lógica de validação
-- Recebe: uma match, a lista de palavras do portuguÊs, uma string de input
-- Retorna: um a tupla com: (True se a coordenada e posicionamento forem válidos,
--                            A string de palavras inválidas,
--                            A quantidade de pontos obtidos,
--                            String de letras inválidas,
--                            String de letras usadas pelo jogador)
initialValidation :: Match -> [String] -> String -> (Bool, [String], Int, [Char], [Char])
initialValidation _ _ "" = (False, [], 0, [], [])
initialValidation match wordlist linha
    | length palavras /= 3 = (False, [], 0, [], []) -- Input completamente errado
    | otherwise = (resCoordenadas, resPalavras, getPointsWord letrasNoBoard word, invalidLetters, letrasUsadas)
    where
        (x, y, isHorizontal, word) = readWordInput linha match
        palavras = words $ map toUpper linha
        coord = (head palavras)
        playerOnTurn = getPlayerOnTurn match
        playerLetters = [letter l | l <- pLetters playerOnTurn]
        letrasNoBoard = _takeUpTo isHorizontal match (x,y) (length word)
        letterUsageReport = _wordLetterReport playerLetters letrasNoBoard word
        invalidLetters = _lettersMissing word letterUsageReport
        letrasUsadas = [l | l <- letterUsageReport, l /= '!', l /= ' ']
        estaConectado = 0 /= (length [x | x <- letrasNoBoard, x `elem` ['A'..'Z']])
        centroLivre = (((curTiles (mBoard match)) !! 7) !! 7) == '-'
        resCoordenadas = (((palavras !! 1) `elem` ["V", "H"]) 
                            && (_coordValidation coord) 
                            && (_wordHasOnlyLettersAndAddsNewLetters word letrasNoBoard) 
                            && (_wordFitsInSpace isHorizontal (x, y) word) 
                            && (_tileValidationDifferentLettersDontOverlap letrasNoBoard word) 
                            && (estaConectado || centroLivre) 
                            && (_coordCenterValidation match isHorizontal (x,y) word))
        resPalavras = (_allWordsExist match wordlist (getWords (placeWord (x,y,isHorizontal,word) (mBoard match))))


-- Recebe: uma match
-- Retorna: o jogador da vez
getPlayerOnTurn :: Match -> Player
getPlayerOnTurn match
    | mTurn match = mP2 match
    | otherwise = mP1 match


-- Recebe: uma match, o booleano que informa se é horizontal, as coordenadas x e y, a palavra
-- Retorna: True se for a primeira palavra deve ocupar o centro, False se não
_coordCenterValidation :: Match -> Bool -> (Int, Int) -> String -> Bool
_coordCenterValidation match isHorizontal (x,y) word
    | centerValue /= '-' = True -- Centro já estava ocupado: não é a primeira palavra
    | isHorizontal = (y == 7) && (x <= 7 && 7 <= finalXInd)
    | otherwise = (x == 7) && (y <= 7 && 7 <= finalYInd)
    where
        finalXInd = x + (length word) - 1
        finalYInd = y + (length word) - 1
        matrix = curTiles (mBoard match)
        centerValue = (matrix !! 7) !! 7


-- Recebe: uma match, a lista de palavras do português, a palvra a ser testada
-- Retorna: True se uma palavra está no arquivo de palavras, False se não
_wordExistenceValidation :: Match -> [String] -> String -> Bool
_wordExistenceValidation match wordList word = ((map toLower word) `elem` (mUsedWords match)) || ((map toLower word) `elem` wordList)


-- Verifica que todas as palavras formadas no tabuleiro existem
-- Recebe: uma match, a lista de todas as palavras do português, a lista de palavras a testar
_allWordsExist :: Match -> [String] -> [String] -> [String]
_allWordsExist match wordlist wordsToTest = [x | x <- wordsToTest, not (_wordExistenceValidation match wordlist x)]


-- Verifica que a coordenada passada é válida: o primeiro caracter é uma letra de A a O e a segunda é uma string com um int de 0 a 14
-- Recebe: "coord", a primeira palavra do input
_coordValidation :: [Char] -> Bool
_coordValidation (x:y) = (x `elem` ['A' .. 'O']) && (UT.isStringInt y) && ((read y :: Int) >= 0 && (read y :: Int) <= 14)


-- Verifica que a palavra jogada pelo jogador é composta apenas de letras e que ela adiciona alguma letra nova ao board
-- Recebe: palavra, letras na posição que ela vai ocupar
-- Retorna: se é uma posição válida
_wordHasOnlyLettersAndAddsNewLetters :: String -> String -> Bool
_wordHasOnlyLettersAndAddsNewLetters word letrasNoBoard = ([] == [l | l <- word, not (isLetter l)]) && (length word /= length [l | l <- letrasNoBoard, isLetter l])


-- Verifica que não haverá sobreposição de letras: o jogador não vai colocar uma letra diferente em um espaço já ocupado por outra
-- Recebe: as letras/simbolos do board na posição que será ocupada, a palavra
-- Retorna: falso se a uma letra das tiles for diferente de uma letra na palavra na mesma posição
_tileValidationDifferentLettersDontOverlap :: String -> String -> Bool
_tileValidationDifferentLettersDontOverlap [] [] = True
_tileValidationDifferentLettersDontOverlap (tileHead:tileTail) (wordHead:wordTail)
    | (tileHead `elem` ['A'..'Z']) && (tileHead /= wordHead) = False
    | otherwise = _tileValidationDifferentLettersDontOverlap tileTail wordTail


-- Recebe: o booleano que informa se eh horizontal, as coordenadas x e y (col, row) indexadas em zero, e a palavra
-- Retorna: True se a palavra cabe no espaço disponível no board, False se não
_wordFitsInSpace :: Bool -> (Int, Int) -> String -> Bool
_wordFitsInSpace isHorizontal (x, y) word
    | isHorizontal = (x <= 15 - (length word) && x >= 0) && (y >= 0 && y <= 14)
    | otherwise = (y <= 15 - (length word) && y >= 0) && (x >= 0 && x <= 14)


-- A função gera um relatório que traz informações sobre as letras usadas
-- Recebe: letras (chars) do jogador, tiles do board na posição que será ocupada, palavra que será colocada
-- Retorna: [Char] que traz informações sobre quais letras do player foram consumidas, quais estavam no board, e quais o jogador não tinha
{-  
    Exemplo de output:

    playerLetters = [A C E <] inicialmente

    PALAVRA: [A B C D E F] tinha A e D no board, não usou do player -> não tinha B e o player não tinha mas tinha curinga, usa o curinga -> ...
   NO BOARD: [A - - D - -] ... não tinha C e E o player tinha, usa C e E -> Não tinha F no board nem o player tinha curinga, vai exclamação
     OUTPUT: [  < C   E !]

     playerLetters = [A] final

     Assim, para saber o que o player usou olha os caracteres e pra saber se passou na validação olha se tem exclamação
     As posições estão sendo respeitadas, então pra saber qual letra faltou basta olhar qual é a letra da palavra na posição exclamação.
 -}
_wordLetterReport :: [Char] -> [Char] -> [Char] -> [Char]
_wordLetterReport _ _ [] = []
_wordLetterReport playerLetters (t:tileTail) (w:wordTail) -- (letrasNoBoard) (word)
    | w == t = ' ':(_wordLetterReport playerLetters tileTail wordTail)
    | w `elem` playerLetters = w:(_wordLetterReport (_removeChar w playerLetters) tileTail wordTail)
    | '<' `elem` playerLetters = '<':(_wordLetterReport (_removeChar '<' playerLetters) tileTail wordTail)
    | otherwise = '!':(_wordLetterReport playerLetters tileTail wordTail)


-- Recebe: uma Match e uma letter, obtém o jogador a partir da match
-- Retorna: True se o jogador tem a Letter passada, false se não 
playerHasLetter :: Match -> Letter -> Bool
playerHasLetter match letter = letter `elem` (pLetters player)
    where player = getPlayerOnTurn match


-- Recebe: palavra a ser testada, string do _wordLetterReport
-- Retorna: as letras que o jogador não tinha para completar a palavra
_lettersMissing :: [Char] -> [Char] -> [Char]
_lettersMissing word filterString =
    map fst $ filter (\(_, el) -> el == '!') $ zip word filterString


-- Recebe: um input em string, uma match
-- Retorna: extrai as informações necessárias para validar o input e retorna as informações contidas na string
readWordInput :: String -> Match -> (Int, Int, Bool, String)
readWordInput linha match = (x, y, isHorizontal, word)
    where
        palavras = words $ map toUpper linha
        coord = (head palavras)
        isHorizontal = (palavras !! 1) == "H"
        word = (palavras !! 2)
        x = ord (head coord ) - ord 'A'
        y = (read (tail coord) :: Int)
        letrasNoBoard = _takeUpTo isHorizontal match (x,y) (length word)


-- Verifica se uma conta existe baseado no nome
-- Recebe: nome da conta
-- Retorna: True se a conta existir, False se não
accExistsValidation :: String -> IO(Bool)
accExistsValidation accName = do
    acc <- getAccByName accName
    case acc of
        Just acc -> return True
        Nothing -> return False


-- Verifica se uma conta existe baseado no nome
-- Recebe: nome da conta
-- Retorna: True se a conta existir, False se não
matchExistsValidation :: String -> IO(Bool)
matchExistsValidation matchName = do
    match <- getMatchByName matchName
    case match of
        Just match -> return True
        Nothing    -> return False


-- Obtém as tiles presentes no board no intervalo especificado
-- Recebe: o booleano se informa se é horizontal, uma match, coordenadas x e y, o tamanho do intervalo
-- Retorna: as tiles no intervalo especificado
_takeUpTo :: Bool -> Match -> (Int, Int) -> Int -> [Char]
_takeUpTo isHorizontal match (x, y) len
    | isHorizontal = (take len (drop x (b !! y)))
    | otherwise = take len $ map (!! x) $ drop y b
    where b = curTiles (mBoard match)

File: prolog/Utils/validator.pl
report(0, Report) :- Report = [false, 0, [], [], []].
report(1, Points, ValidLetters, InvalidLetters, InvalidWords, Report) :- Report = [true, Points, ValidLetters, InvalidLetters, InvalidWords], !.


% Encapsula lógica de validação
% Recebe: o nome de uma match
% Recebe: uma string de input
% Retorna: uma lista com: [True se a coordenada e posicionamento forem válidos,
%                            A quantidade de pontos obtidos,
%                            Uma lista de letras usadas pelo jogador,
%                            Uma lista de letras que o jogador não possui,
%                            Uma lista de palavra inválidas]
validation(MatchName, InputLine, Report) :-
    get_match_board_name(MatchName, BoardName),

    (read_input(InputLine, Info) ->

        nth0(0, Info, X),
        nth0(1, Info, Y),
        nth0(2, Info, WordLetters),
        nth0(3, Info, IsHorizontal),

        % Lógica que verifica se o jogador tem as letras da palavra
        get_turn_player_name(MatchName, PlayerName),
        get_player_letters(MatchName, PlayerName, PlayerLetters),
        length(WordLetters, WordLength),
        
        (IsHorizontal -> N is X + WordLength ; N is Y + WordLength), get_work_tiles(BoardName, WorkTiles),

        take_up_to(WorkTiles, X, Y, N, IsHorizontal, BoardTiles), 
        get_element(WorkTiles, 7, 7, MiddleElement),
        (not(maplist(special_char, BoardTiles)) ; special_char(MiddleElement)),
        player_has_letters(WordLetters, PlayerLetters, BoardTiles, ValidLetters, InvalidLetters),
        not(maplist(is_alpha, BoardTiles)), letter_overlap_validation(WordLetters, BoardTiles), 
        get_points_word(BoardTiles, WordLetters, Points),

        % Lógica de validação da palavra

        ((word_fits_in_space(X, Y, WordLetters, IsHorizontal)) ->
            (center_tile_validation(WorkTiles, X, Y, IsHorizontal, WordLetters) ->
                atomic_list_concat(WordLetters, Word),
                place_word(X, Y, IsHorizontal, Word, BoardName, NewBoard),
                get_words(NewBoard, BoardWords),

                (all_words_exist(BoardWords, InvalidWords) ->
                    report(1, Points, ValidLetters, InvalidLetters, InvalidWords, Report),
                    update_work_tiles(BoardName, NewBoard),!
                )
                
            )
        )
    )

    ;

    report(0, Report).

% Encapsula o tratamento da linha de entrada do jogador
% Recebe a string de entrada
% Retorna: uma lista com as informações tratadas: [Coordenada X, Coordenada Y, 
%                                                   Uma lista com os caracteres da palavra,
%                                                   True se a direção é horizontal]
read_input(InputLine, Info) :-
    string_upper(InputLine, InputLineUpper),
    split_string(InputLineUpper, " ", "", Input), length(Input, 3),
    nth0(0, Input, InputCoord),
    nth0(1, Input, Direction), (Direction = "H", ! ; Direction = "V"),
    nth0(2, Input, InputWord), string_chars(InputWord, WordLetters), 
    length(WordLetters, Len), Len >= 2, maplist(is_alpha, WordLetters),

    string_chars(InputCoord, [H|T]), length([H|T], 3),
    char_code(H, CharCode), 
    nth0(0, T, A1), char_type(A1, digit),
    nth0(1, T, A2), char_type(A2, digit),
    atom_concat(A1, A2, Concat),
    atom_string(Concat, NewCoord),
    
    X is CharCode - 65, (X >= 0, X =< 14),
    number_codes(Y, NewCoord), (Y >= 0, Y =< 14),

    (Direction = "H" ->
        IsHorizontal = true
    ; IsHorizontal = false),

    Info = [X, Y, WordLetters, IsHorizontal].

is_alpha(Char) :-
    char_type(Char, alpha).

% Verifica se o jogador tem as letras da palavra jogada
% Recebe: A lista com os caracteres da palavra
% Recebe: As letras do jogador
% Recebe: Uma sublista do espaço do tabuleiro em que a palavra possivelmente será inserida
% Retorna: Uma lista de letras válidas utilizadas
% Retorna: Uma lista de letras inválidas
player_has_letters([], _, _, [], []).
player_has_letters([WL|WLs], PlayerLetters, [WL|BTs], ValidLetters, InvalidLetters) :-
    player_has_letters(WLs, PlayerLetters, BTs, ValidLetters, InvalidLetters), !.
player_has_letters([WL|WLs], PlayerLetters, [_|BTs], ValidLetters, InvalidLetters) :-
    (member(WL, PlayerLetters) ->
        select(WL, PlayerLetters, NewPlayerLetters),
        player_has_letters(WLs, NewPlayerLetters, BTs, TempValidLetters, InvalidLetters),
        ValidLetters = [WL|TempValidLetters], !
    ; member('<', PlayerLetters) ->
        select('<', PlayerLetters, NewPlayerLetters),
        player_has_letters(WLs, NewPlayerLetters, BTs, TempValidLetters, InvalidLetters),
        ValidLetters = ['<'|TempValidLetters]), !.
player_has_letters([WL|WLs], _, [_|BTs], ValidLetters, InvalidLetters) :-
    player_has_letters(WLs, _, BTs, ValidLetters, TempInvalidLetters),
    InvalidLetters = [WL|TempInvalidLetters].

% Verifica se o a palavra cabe na coordenada passada
% Recebe: Coordenada X
% Recebe: Coordenada Y
% Recebe: Lista de caracteres da palavra
% Recebe: Booleano para a direção horizontal
word_fits_in_space(X, Y, WordLetters, IsHorizontal) :-
    length(WordLetters, WordLength),

    (IsHorizontal ->
        (X =< 15 - WordLength)
    ; (Y =< 15 - WordLength)).

% Verifica se a sobreposição da palavra sobre o tabuleiro é válida
% Recebe: Lista com os caracteres da palavra
% Recebe: Sublista do espaço do tabuleiro em que a palavra possivelmente será inserida
letter_overlap_validation([], []) :- !.
letter_overlap_validation([H|T], [H|Y]) :-
    letter_overlap_validation(T, Y), !.
letter_overlap_validation([_|T], [X|Y]) :-
    not(is_alpha(X)),
    letter_overlap_validation(T, Y), !.

% Obtém as tiles presentes no board no intervalo especificado
% Recebe: uma matriz
% Recebe: coordenadas x e y
% Recebe: o tamanho do intervalo
% Recebe: o booleano se informa se é horizontal
% Retorna: as tiles no intervalo especificado
take_up_to(_, N, _, N, true, []) :- !.
take_up_to(_, _, N, N, false, []) :- !.
take_up_to(Matrix, X, Y, N, IsHorizontal, BoardTiles) :-
    nth0(Y, Matrix, List),
    nth0(X, List, Char),

    (IsHorizontal ->
        NewX is X + 1,
        take_up_to(Matrix, NewX, Y, N, IsHorizontal, TempBoardTiles)
    ;
        NewY is Y + 1,
        take_up_to(Matrix, X, NewY, N, IsHorizontal, TempBoardTiles)),

    append([Char], TempBoardTiles, BoardTiles).

% Verifica se a primeira jogada utilizou a tile central
% Recebe: uma matriz
% Recebe: s coordenadas X e Y
% Recebe: booleano para a direção horizontal
% Recebe: lista de caracteres da palavra
center_tile_validation(WorkTiles, X, Y, IsHorizontal, WordLetters) :-
    take_up_to(WorkTiles, 7, 7, 8, IsHorizontal, [CenterTile]),
    length(WordLetters, WordLength),

    ((is_alpha(CenterTile), !) ;

    (IsHorizontal ->
        WordLastInd is X + WordLength - 1,
        Y =:= 7, X =< 7, WordLastInd >= 7
    ; WordLastInd is Y + WordLength - 1,
        X =:= 7, Y =< 7, WordLastInd >= 7)).

% Verifica se todas as palavras formas no tabuleiro existem na língua portuguesa
% Recebe: Lista de palavras do tabuleiro
% Retorna: Lista de palavras inválidas
all_words_exist([], []).
all_words_exist([H|T], InvalidWords) :-
    all_words(AW),
    downcase_atom(H, Lowercase),
    (memberchk(Lowercase,AW) ->
        all_words_exist(T, InvalidWords), !
    ;
        all_words_exist(T, TempInvalidWords),
        InvalidWords = [H|TempInvalidWords]), !.

get_sublist_col(_, To, To, _, []) :- !.
get_sublist_col(Matrix, From, To, X, Sublist) :-
    nth0(From, Matrix, Col),
    nth0(X, Col, Char),
    NewFrom is From + 1,
    get_sublist_col(Matrix, NewFrom, To, X, TempSublist),
    append([Char], TempSublist, Sublist).

valid_name(Name) :-
    \+ Name = "",
    \+ contains_space(Name),
    \+ contains_special_chars_helper(Name).

contains_space(Name) :-
    sub_string(Name, _, _, _, ' ').

contains_special_chars(Name) :-
    writeln("Checking for special characters"),
    atom_chars(Name, Chars),
    contains_special_chars_helper(Chars).

contains_special_chars_helper([]).
contains_special_chars_helper([Char|_]) :-
    special_char(Char),
    !.
contains_special_chars_helper([_|Rest]) :-
    contains_special_chars_helper(Rest).

special_char(Char) :-
    member(Char, ['!', '@', '#', '~', '$', '%', '^', '&', '*', '(', ')', '-', '_', '=', '+', '[', ']', '{', '}', ';', ':', "'", '"', '<', '>', ',', '.', '/', '?', '\\', '|',
                  'á', 'à', 'ã', 'â', 'é', 'è', 'ẽ', 'ê', 'í', 'ì', 'ĩ', 'î', 'ó', 'ò', 'õ', 'ô', 'ú', 'ù', 'ũ', 'û', 'ç']).
